File: Data Structures & Algorithms/jump-game/submission-6.go

func helper(i int,nums []int, dp []bool) bool{
    if i == 0 {
        dp[i] = true
        return true
    }

    if dp[i] == true{
        return true
    }

    for j := i-1; j>= 0;j-- {
        if nums[j] + j >= i && helper(j,nums, dp) == true {
            dp[i] = true
             break
        }
    }
    return dp[i]
}

func canJump(nums []int) bool {
    n := len(nums)

    dp := make([]bool,n)

    helper(n-1,nums,dp)

    return dp[n-1]

    

}

File: Data Structures & Algorithms/house-robber-ii/submission-6.py
class Solution:
    def helper(self, nums: List[int], p: int, e:int, dp: List[int]) -> int:
        if p >= e:
            return 0

        if dp[p] != -1:
            return dp[p]
        
        pick = nums[p] + self.helper(nums,p+2,e,dp)
        dpick = self.helper(nums,p+1,e,dp)

        dp[p] = max(pick,dpick)

        return dp[p]

    def rob(self, nums: List[int]) -> int:
        l = len(nums)
        dp1 = [-1]*l
        dp2 = [-1]*l
        if l == 1:
            return nums[0]

        return max(self.helper(nums,0,l-1,dp1), self.helper(nums,1,l,dp2))

